File: backend/app/core/security.py
"""Password hashing and JWT issuing/verification."""
from __future__ import annotations

import time
import uuid
from typing import Any

import jwt
from argon2 import PasswordHasher
from argon2.exceptions import VerificationError, VerifyMismatchError

from app.core.config import settings
from app.core.errors import AuthenticationError

_hasher = PasswordHasher()


def hash_password(password: str) -> str:
    return _hasher.hash(password)


def verify_password(password: str, hashed: str) -> bool:
    try:
        return _hasher.verify(hashed, password)
    except (VerifyMismatchError, VerificationError):
        return False


def _encode(subject: str, org_id: str, role: str, ttl: int, token_type: str) -> str:
    now = int(time.time())
    payload = {
        "sub": subject,
        "org": org_id,
        "role": role,
        "type": token_type,
        "iat": now,
        "exp": now + ttl,
        "jti": uuid.uuid4().hex,
    }
    return jwt.encode(payload, settings.jwt_secret, algorithm=settings.jwt_algorithm)


def create_access_token(subject: str, org_id: str, role: str) -> str:
    return _encode(subject, org_id, role, settings.jwt_access_ttl_seconds, "access")


def create_refresh_token(subject: str, org_id: str, role: str) -> str:
    return _encode(subject, org_id, role, settings.jwt_refresh_ttl_seconds, "refresh")


def decode_token(token: str, expected_type: str = "access") -> dict[str, Any]:
    try:
        payload = jwt.decode(token, settings.jwt_secret, algorithms=[settings.jwt_algorithm])
    except jwt.PyJWTError as exc:  # expired, malformed, or bad signature
        raise AuthenticationError("Token is invalid or expired.") from exc
    if payload.get("type") != expected_type:
        raise AuthenticationError("Wrong token type.")
    return payload

File: apps/web/lib/server/auth/security.ts
/**
 * Password hashing and JWT issuing/verification. Port of backend/app/core/security.py.
 *
 * Argon2 (not bcrypt) to match the Python side's hashing scheme, via `@node-rs/argon2`
 * — a napi-rs package with prebuilt binaries per platform (including Vercel's Node
 * runtime), avoiding the node-gyp native-compile risk a plain `argon2` package would
 * carry there.
 */
import { hash, verify } from "@node-rs/argon2";
import { SignJWT, jwtVerify, type JWTPayload } from "jose";
import { randomUUID } from "crypto";
import { settings } from "@/lib/server/config";
import { AuthenticationError } from "@/lib/server/errors";

export async function hashPassword(password: string): Promise<string> {
  return hash(password);
}

export async function verifyPassword(password: string, hashed: string): Promise<boolean> {
  try {
    return await verify(hashed, password);
  } catch {
    return false;
  }
}

export interface AccessClaims extends JWTPayload {
  sub: string;
  org: string;
  type: "access" | "refresh";
}

const secret = () => new TextEncoder().encode(settings.jwtSecret);

async function encode(subject: string, orgId: string, ttlSeconds: number, type: "access" | "refresh") {
  const now = Math.floor(Date.now() / 1000);
  return new SignJWT({ sub: subject, org: orgId, type, jti: randomUUID().replace(/-/g, "") })
    .setProtectedHeader({ alg: settings.jwtAlgorithm })
    .setIssuedAt(now)
    .setExpirationTime(now + ttlSeconds)
    .sign(secret());
}

export function createAccessToken(subject: string, orgId: string): Promise<string> {
  return encode(subject, orgId, settings.jwtAccessTtlSeconds, "access");
}

export function createRefreshToken(subject: string, orgId: string): Promise<string> {
  return encode(subject, orgId, settings.jwtRefreshTtlSeconds, "refresh");
}

export async function decodeToken(
  token: string,
  expectedType: "access" | "refresh" = "access"
): Promise<AccessClaims> {
  let payload: JWTPayload;
  try {
    ({ payload } = await jwtVerify(token, secret(), { algorithms: [settings.jwtAlgorithm] }));
  } catch {
    throw new AuthenticationError("Token is invalid or expired.");
  }
  if (payload.type !== expectedType) throw new AuthenticationError("Wrong token type.");
  return payload as AccessClaims;
}
